### PR TITLE
refactor(header): migrate to fast element

### PR DESCRIPTION
**Pull request by rinaok**
### Changes:
1. Added a `fixed` attribute.
2. Added `heading` attribute instead of `slot`.

### Changes to consider:
1. Change the name to _header_ / _pageHeader_ / _globalHeader_

### CODE REVIEW

**Comment by yinonov** [comment]
can we have an example in **side-drawer** with top-app-bar?

File: libs/components/src/lib/top-app-bar/README.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        background-color: var(--vvd-color-neutral-10);
+        margin:0;
+    }
+</style>
+<vwc-top-app-bar heading="Alternate Top App Bar" alternate></vwc-top-app-bar>
+```
+
+## Slots
+
+### Meta
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:77]: I'm not sure that's its right name

File: libs/components/src/lib/top-app-bar/top-app-bar.scss [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	width: 100%;
+	height: var(#{variables.$top-app-bar-height});
+	box-sizing: border-box;
+	justify-content: space-between;
+	padding: 8px 12px;
+	background-color: var(#{theme-consts.$vvd-color-canvas});
+	color: var(#{theme-consts.$vvd-color-on-canvas});
+	column-gap: 12px;
+	
+	&.shadow {
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/top-app-bar.scss:26]: elevated?

File: libs/components/src/lib/top-app-bar/README.md [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+## Properties
+
+### Heading
+Add the `heading` attribute to set the heading text.
+
+- Type: `string`
+- Default: `undefined`
+
+```html preview full
+<style>
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:18]: don't see how this helps the demo. unecessary decoration

File: libs/components/src/lib/top-app-bar/README.md [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+### Fixed
+Add the `fixed` attribute to set the top-app-bar's position to be fixed.
+
+- Type: `boolean`
+- Default: `false`
+
+```html preview full
+<style>
+    body {
+        background-color: var(--vvd-color-neutral-10);
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:35]: again, unecessary

File: libs/components/src/lib/top-app-bar/README.md [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+<vwc-top-app-bar heading="Alternate Top App Bar" alternate></vwc-top-app-bar>
+```
+
+## Slots
+
+### Meta
+The `meta` slot is for action content next to the heading.
+
+```js
+<vwc-button slot="meta" icon="menu-line"></vwc-button>
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:63]: what is that for? it's no js

File: libs/components/src/lib/top-app-bar/README.md (hunk 3 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:18]: Screen Shot 2022-05-29 at 16 53 09

without it, it looks like this

File: libs/components/src/lib/top-app-bar/README.md (hunk 4 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:35]: Screen Shot 2022-05-29 at 16 53 43

with the background you can understand where is the top bar

File: libs/components/src/lib/top-app-bar/README.md (hunk 5 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:63]: it's an example for a slot. Should it be html? I copied it from the side-drawer: https://vivid.deno.dev/components/side-drawer#app-content
It's `js` there too

File: libs/components/src/lib/top-app-bar/README.md (hunk 1 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:59]: what is the right name?

File: libs/components/src/lib/top-app-bar/README.md [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    <vwc-side-drawer id="sideDrawer" open slot="app-content">
+        <div>
+            <vwc-sidenav-item href="#" text="1st level item" icon="home-line"></vwc-sidenav-item>
+            <p><vwc-text font-face="body-2-bold">SECTION TITLE</vwc-text></p>
+            <vwc-sidenav-item href="#" text="1st level item" icon="chat-line"></vwc-sidenav-item>
+            <vwc-sidenav-item href="#" text="1st level item" icon="chat-line"></vwc-sidenav-item>
+        </div>
+        <div slot="app-content">
+            <vwc-button id="button" shape="pill" icon='menu-solid'></vwc-button>
+            <vwc-text font-face="body-1">
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:180]: wrap with layout? it's a usage example so decoration is relevant

File: libs/components/src/lib/top-app-bar/README.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+        <vwc-button icon="heart-solid"></vwc-button>
+    </div>
+</vwc-top-app-bar>
+```
+
+### App-content
+
+To add content aside the side-drawer, add a slot called `app-content`.
+
+```js
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:109]: why the duplication of source? and this ain't no JS

**Comment by rinaok** [comment]
@yinonov There is only one thing left to do: choose a name instead of top-app-bar.

**Review by yinonov — CHANGES_REQUESTED** [review]
that effect is quite unbelievable 😁
image

File: libs/components/src/lib/top-app-bar/README.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+### Z-index
+
+Controls the z-index of the top-app-bar.
+
+- `--top-app-bar-z-index`
+- Type: `Number`
+- Default: `7`
+
+### Height
```
Comment (yinonov) [inline libs/components/src/lib/top-app-bar/README.md:203]:
```suggestion
### Block Size
```

**Review by yinonov — CHANGES_REQUESTED** [review]

File: libs/components/src/lib/top-app-bar/README.md (hunk 8 above)
Comment (yinonov) [reply libs/components/src/lib/top-app-bar/README.md:203]: is it by spec to expose block size value as a wild card?

**Comment by rinaok** [comment]
> a scrolled app-header (non fixed) should scroll back instantly when page scrolls back down. start by applying the following
> 
> ```css
> position: sticky;
> top: calc(-1 * var(--top-app-bar-height));
> ```

We've already tried that. @rachelbt , can you elaborate?

File: libs/components/src/lib/top-app-bar/README.md (hunk 8 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:203]: @jshenkman Should we expose the Top Bar's height to the user?

Comment (yinonov) [reply libs/components/src/lib/top-app-bar/README.md:203]: I think it should be a `min-block-size` only. if content exceeds, the top-bar extends. naturally

File: libs/components/src/lib/top-app-bar/README.md (hunk 1 above)
Comment (yinonov) [reply libs/components/src/lib/top-app-bar/README.md:96]: can you provide a few from other libraries usages?

**Comment by rinaok** [comment]
> that effect is quite unbelievable 😁 image

@yinonov Does it look good now?

File: libs/components/src/lib/top-app-bar/README.md (hunk 8 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:203]: @yinonov @rachelbt  The `top-app-bar-height` effects also on the `padding-top` of the `app-content`.
So how can I do it when the content exceeds?

File: libs/components/src/lib/top-app-bar/README.md (hunk 1 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md:96]: @yinonov material calls this slot: **navigationIcon**

File: libs/components/src/lib/top-app-bar/README.md (hunk 8 above)
Comment (rachelbt) [reply libs/components/src/lib/top-app-bar/README.md:203]: if we are using padding - we can't use the min-height. We need a predefined size either by us or by the user

**Comment by rachelbt** [comment]
> > a scrolled app-header (non fixed) should scroll back instantly when page scrolls back down. start by applying the following
> > ```css
> > position: sticky;
> > top: calc(-1 * var(--top-app-bar-height));
> > ```
> 
> We've already tried that. @rachelbt , can you elaborate?

I think its the mixture of shadow & light dom that makes sticky not properly work.
It can work id we "fix" the top-app-bar host to the screen with:
```
position: fixed;
inset: 0;
overflow-y: auto;
```
But I'm not sure this is a good idea...

File: libs/components/src/lib/top-app-bar/README.md (hunk 1 above)
Comment (yinonov) [reply libs/components/src/lib/top-app-bar/README.md]: right but it's often not what it contains

File: libs/components/src/lib/top-app-bar/README.md (hunk 8 above)
Comment (rinaok) [reply libs/components/src/lib/top-app-bar/README.md]: @yinonov It's not related to this PR, it should be fixed in a new PR cause it happens in all the components.


Screen Shot 2022-06-09 at 12 28 43

File: libs/components/src/lib/header/header.spec.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+			window['pageYOffset'] = scrollDist;
+			window.dispatchEvent(new CustomEvent('scroll'));
+			await elementUpdated(element);
+		}
+
+		beforeEach(function () {
+			element.fixed = true;
+		});
+
+		it('should default to no elevation with no scrolling', async function () {
```
Comment (YonatanKra) [inline libs/components/src/lib/header/header.spec.ts:76]: I can't see any relation to scrolling in the test itself.
Maybe this is more accurate:
```suggestion
		it('should default to no elevation', async function () {
```

File: libs/components/src/lib/header/header.spec.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+			await elementUpdated(element);
+			const control = getControlElement(element);
+			expect(element.elevated).toEqual(false);
+			expect(control.classList.contains('elevated')).toEqual(false);
+		});
+
+		it('should set "elevated" to true and add "elevated" class', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/header/header.spec.ts:83]:
```suggestion
		it('should set "elevated" to true and add "elevated" class when scrolled', async () => {
```

File: libs/components/src/lib/header/header.spec.ts [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+		it('should remove elevated state when not fixed', async function () {
+			element.fixed = false;
+			const control = getControlElement(element);
+			await emulateScroll();
+			const hasClassElevated = control.classList.contains('elevated');
+
+			expect(hasClassElevated).toEqual(false);
+		});
+
+		it('should remove elevated state when offset is 0', async function () {
```
Comment (YonatanKra) [inline libs/components/src/lib/header/header.spec.ts:103]:
```suggestion
		it('should remove elevated state when scroll offset is 0', async function () {
```

File: libs/components/src/lib/header/README.md [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+        </p>
+      </vwc-text>
+    </vwc-layout>
+  </div>
+</vwc-header>
+```
+
+## CSS Custom Properties
+
+### Z-index
```
Comment (yinonov) [inline libs/components/src/lib/header/README.md:230]:
```suggestion
### Stacking Context
```

File: libs/components/src/lib/header/README.md [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+
+Controls the z-index of the header.
```
Comment (yinonov) [inline libs/components/src/lib/header/README.md:232]: _"use the `--header-z-index` ..."_

File: libs/components/src/lib/header/README.md [hunk 14]
```diff
@@ -0,0 +1,236 @@
...
+
+- `--header-z-index`
```
Comment (yinonov) [inline libs/components/src/lib/header/README.md:234]:
```suggestion
```

File: libs/components/src/lib/header/header.template.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+	FoundationElementDefinition,
+} from '@microsoft/fast-foundation';
+import { classNames } from '@microsoft/fast-web-utilities';
+import type { header } from './header';
+
+const getClasses = ({ fixed, alternate, elevated }: header) => classNames(
+	'control',
+	['fixed', fixed],
+	['alternate', alternate],
+	['elevated', elevated],
```
Comment (yinonov) [inline libs/components/src/lib/header/header.template.ts:14]:
```suggestion
	['fixed', Boolean(fixed)],
	['alternate', Boolean(alternate)],
	['elevated', Boolean(elevated)],
```

File: libs/components/src/lib/header/header.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+ */
+export class header extends FoundationElement {
+	/**
+	 *
+	 *headerEl
+	 * @public
+	 *
+	 * HTML Attribute: heading
+	 */
+	@attr heading?: string;
```
Comment (yinonov) [inline libs/components/src/lib/header/header.ts:17]: is this defined / required by design?

File: libs/components/src/lib/header/header.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+	 * sets the header to be fixed
+	 *
+	 * @public
+	 */
+	@attr({
+		mode: 'boolean',
+	}) fixed = false;
+
+	/**
+	 * applies scheme alternate region
```
Comment (yinonov) [inline libs/components/src/lib/header/header.ts:29]:
```suggestion
	 * applies scheme alternate to header region
```

File: libs/components/src/lib/header/header.ts [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+		window.addEventListener('scroll', this.#updateElevatedState);
+	}
+
+	override disconnectedCallback() {
+		super.disconnectedCallback();
+		window.removeEventListener('scroll', this.#updateElevatedState);
+	}
+
+	#updateElevatedState = () => {
+		this._elevated = !this._elevated;
```
Comment (yinonov) [inline libs/components/src/lib/header/header.ts:50]: if not used in template then do `#elevated`

File: libs/components/src/lib/header/header.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+	 */
+	@attr({
+		mode: 'boolean',
+	}) alternate = false;
+
+	@observable _elevated = false;
+
+	override connectedCallback() {
+		super.connectedCallback();
+		window.addEventListener('scroll', this.#updateElevatedState);
```
Comment (yinonov) [inline libs/components/src/lib/header/header.ts:41]: this type of code will not comply to an SSR practice in future plans.
not saying no as there might not be another option. but asking, is there another option that will not throw in node run time?

File: libs/components/src/lib/header/index.ts [hunk 20]
```diff
@@ -0,0 +1,14 @@
+import type { FoundationElementDefinition } from '@microsoft/fast-foundation';
+import { designSystem } from '../../shared/design-system';
+import styles from './header.scss';
+
+import { header } from './header';
+import { headerTemplate as template } from './header.template';
+
+export const vividheader = header.compose<FoundationElementDefinition>({
```
Comment (yinonov) [inline libs/components/src/lib/header/index.ts:8]:
```suggestion
export const vividHeader = Header.compose<FoundationElementDefinition>({
```

**Review by yinonov — CHANGES_REQUESTED** [review]

File: libs/components/src/lib/header/README.md (hunk 12 above)
Comment (rachelbt) [reply libs/components/src/lib/header/README.md:230]: I think this should come before the usage examples

Comment (rinaok) [reply libs/components/src/lib/header/README.md:230]: @yinonov Why z-index is not a good title?
Side-Drawer has this property: 

Screen Shot 2022-06-22 at 12 10 42

File: libs/components/src/lib/header/header.ts (hunk 18 above)
Comment (rinaok) [reply libs/components/src/lib/header/header.ts:50]: @yinonov I can't make an `@observable ` to be a private member.

File: libs/components/src/lib/header/header.ts (hunk 16 above)
Comment (rinaok) [reply libs/components/src/lib/header/header.ts:17]: @yinonov Card, Note, Calendar and Accordion have this attribute so I think that it should be consistent.

File: libs/components/src/lib/header/header.ts (hunk 19 above)
Comment (rinaok) [reply libs/components/src/lib/header/header.ts:41]: @yinonov How can I debug it?

File: libs/components/src/lib/header/README.md (hunk 12 above)
Comment (yinonov) [reply libs/components/src/lib/header/README.md:230]: it's not bad, thought the suggestion is better but if doesn't ring right, resolve the comment

File: libs/components/src/lib/header/header.ts (hunk 16 above)
Comment (yinonov) [reply libs/components/src/lib/header/header.ts:17]: I think we should reconsider the way we define top text hierarchy. given that _heading_ should probably introduce an actual heading element (Hx) in template.

certain cases, like note, where Hx isn't wrapping the text, title might be a better member name

image

@rachelbt FYI. should we open this up?

File: libs/components/src/lib/side-drawer/side-drawer.scss [hunk 21]
```diff
@@ -48,12 +48,14 @@ $inline-size: var(#{variables.$side-drawer-inline-size}, 280px);
...
-			::slotted(vwc-top-app-bar),
-			::slotted(vwc-top-app-bar-fixed) {
-				--mdc-top-app-bar-width: calc(100% - #{$inline-size});
-			}
+
+			// TODO: check
+			// @rinaok - this is a hack to get the drawer to be the same width as the app
+			// ::slotted(vwc-header) {
+			// 	--mdc-header-width: calc(100% - #{$inline-size});
+			// }
```
[the remark this answers is not in the document]
Comment (yinonov) [reply libs/components/src/lib/side-drawer/side-drawer.scss:58]: no

File: libs/components/src/lib/header/header.ts (hunk 16 above)
Comment (rinaok) [reply libs/components/src/lib/header/header.ts:17]: @yinonov @rachelbt  Maybe you should create a new PR for that and change everything there.

File: libs/components/src/lib/header/ui.test.ts [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+import {
+	extractHTMLBlocksFromReadme,
+	loadComponents,
+	loadTemplate,
+} from '../../visual-tests/visual-tests-utils.ts';
+
+const components = ['header', 'button', 'text', 'side-drawer', 'sidenav-item', 'layout' ];
+
+test('should show the component', async ({ page }: { page: Page }) => {
+	const template = extractHTMLBlocksFromReadme(
```
[the remark this answers is not in the document]
Comment (rinaok) [reply libs/components/src/lib/header/ui.test.ts:13]: @YonatanKra But what can I check there? window's pageYOffset? I check it in the unit tests

File: libs/components/src/lib/header/header.ts (hunk 16 above)
Comment (yinonov) [reply libs/components/src/lib/header/header.ts:17]: https://github.com/Vonage/vivid-3/issues/480

**Review by YonatanKra — APPROVED** [review]

**Pull request merged**